### PR TITLE
RFC: Rewrite PipeStream and SshCommand stream handling

### PR DESCRIPTION
**Pull request by hifi**
+ More efficient when moving large amounts of binary data
    * Speedup on Mono: ~1.71x
    * Speedup on Win10: ~1.39x
 + No in-memory buffering of data
    * Except if you use the Execute() method to return a result
      string
    * The result string will be empty if you have read the output
      stream completely
 - Unless you close or have something reading the error stream it
   can completely block the execution as it will hang the event
   handler indefinitely if no one is reading the other end
 - Reading from PipeStream will now always block until EOF
 - Writing to PipeStream will now always block until a reader has
   completely emptied it

This has not been tested with anything else than SshCommand in a very limited way. I'm planning on adding InputStream on top of this if this refactor is in the right direction.

### CODE REVIEW

**Comment by hifi** [comment]
InputStream is now implemented. It is now possible to fully control the process execution and implement things like git+ssh or rsync+ssh on top of SSH.NET.

**Comment by drieseng** [comment]
Great contributions.
I'll be reviewing them in details asap.
Can you start by adding more unit tests for PipeStream?
I was planning on using a linked list implementation for PipeStream, but that should not affect the behavior (and as such, it should not affect the tests you add).
I'll try to find time for that and based on performance tests, we'll either use that implementation or yours.
OK?

**Comment by hifi** [comment]
I didn't see any reason to *buffer* data in the pipe as that will only increase memory use without improving throughput so my PipeStream implementation keeps only one window of data and expects someone to read it and writing will block until it has been completely read out.

The absolute best way to handle the incoming data from SSH would be to (optionally) throw out the streams and use the events instead as that would completely eliminate read blocking and unnecessary threading/locking that's caused by it.

I'm not 100% sure what the expected behavior of PipeStream is as the non-blocking exception thing was there that I removed and there's a bit of confusing going on with my commits regarding disposing the streams and letting the last buffer to be read out. Flushing was completely removed for instance as I didn't see the need for it in the simple version of PipeStream but it might not be the right move.

Regarding performance I have some doubts a linked list would be any faster as it would increase the amount of work to be done if there's a reader ready to empty it at all times (like `Stream.CopyTo()`) and the list would mostly contain only one element. It would also require you to support partial reading of a written byte buffer which will increase the complexity.

A linked list would also change the behavior from (my) blocking back to non-blocking and it would allow the writer to exhaust the memory if the reader is slow for any reason where usually the speed of transfer is determined by the complete blocking chain.

The original PipeStream implementation resulted in around 50-55% CPU use where my hanged around 20-30%. Throughput during my test was a constant 1Gbit/s with both and OpenSSH hitting local network limit.

**Comment by hifi** [comment]
Streams are supposed to be blocking (right?) so the default being blocking would be something I'd expect. The `Result` property and `EndExecute` method both empty out the stream properly with my PipeStream implementation after shuffling SshCommand a bit so it *should* work even with the blocking behavior.

Making it all not block would be best done by having events as it removes the pipes completely and if no one is listening then the data is simply just lost. Blocking the writer when not using events looks like the right call to me.

The big issue is handling reading both stdout and stderr simultaneously. They can both become large if buffered in-memory and no one is reading them. The simple solution for the user of the api would be to dispose the extended stream immediately to prevent it from blocking if the process writes to stderr and the user does not care about it.

In summary, changing the api of SshCommand and/or adding another command class for blocking use might be worth a shot to support different use cases.

I'm a little bit annoyed when people use the TTY method to run simple commands (with Expect kludge) when SshCommand is usually the right thing to do when you're not running anything interactive so getting it user friendly and performant at the same time is something I have a lot of interest in.

**Comment by hifi** [comment]
I thought about the blocking problem more and I think a simple works-for-everyone solution would be to make SshCommand export the streams as PipeStream instead of plain Stream and have a property for switching between blocking and non-blocking (buffering) mode.

Depending on your performance testing, if a linked list is not significantly slower than the raw byte[] thing I have when there's a reader always available, switching between blocking and non-blocking mode would just add a lock to writer when there's something in the list or alternatively it would switch between a list and a single byte[] buffer if there's a major performance difference.

I can see this benefiting my use case as well as then I could keep OutputStream blocking but let ExtendedOutputStream buffer everything in-memory when I'm not expecting there to be much. It would also make a nice default if all streams buffer as you want them but the property be exposed to switch them on-demand.

What do you think?

**Comment by drieseng** [comment]
I'm still incliding to go for the buffered (non-blocking write, blocking read) approach.
However, give me some time to look into it.
Meanwhile, do you have (more or less selfcontained) integration tests that you use for testing both input and output streams?

**Comment by hifi** [comment]
You can execute 'cat' on the remote system and then loopback everything through it from InputStream to OutputStream. I used some random binary data (files) and hashed them after piping through and measured the CPU time while it was doing that.

**Comment by hifi** [comment]
https://gist.github.com/hifi/9213b8981a938be7fd2dda4790a40dfb an alternative PipeStream that queues everything. It seems to be as heavy as the original implementation (on Mono) but the code is a bit cleaner. Reader is blocking but writer isn't.

**Comment by hifi** [comment]
More testing reveals that the buffering version I gisted does not have almost any performance impact compared to the blocking one I have in this pull request making it the clear winner supporting both workflows.

On Mono, however, the buffering implementation is CPU intensive but that's something that needs to be tackled on that side.

**Comment by hifi** [comment]
Here are some actual benchmarks of the two implementations on three different runtimes: https://github.com/hifi/PipeBench

**Comment by drieseng** [comment]
Let me start by saying: thanks for all the effort you've put into this (getting nothing but semi-negative feedback from me). I don't have much time to spend on this right now (got two branches called **Sofie** and **Lisa** that require attention).

Usage of **BlockingCollection** is a no-go since it's only available as of .NET 4+.
I think performance can be improved (and allocations reduced) by using a linked list approach.
I don't like the fact that the buffer size is limited to the 64 Kb.

Feel free to disregard these "negative" remarks until I have time to work with you on this.

**Comment by hifi** [comment]
Added an implementation of BufferingPipeStream to the benchmark that does its own locking around Queue. It is the fastest implementation of them all on Microsoft runtimes.

**Comment by drieseng** [comment]
Apart from both memory and performance optimizations, one of the most important changes in the new implementation is that ```Read(byte[] buffer, int offset, int count)``` will block until either the request number of bytes are available or the stream is closed.

Before we can merge this new implementation, we should:

- [ ] Add unit tests that fully cover the new implementation it.

- [ ] Review our codebase where we use PipeStream and (may) count on it to only block until at least some data is available.

- [ ] Decide whether to use ```PipeStream.Dispose()``` to mark the stream as EOF, or introduce a separate (internal) method for this purpose.

- [ ] Modify our codebase to use the appropriate mechanism to signal that no more data is available.

- [ ] Fix unit tests that are broken due to the new behavior.

- [ ] Ensure we document the public-facing part of this breaking change for the next release, and provide a clear indication where users may be confronted with this change and how they can or should deal with it.

**Comment by drieseng** [comment]
@hifi What do you think about this:  should we use PipeStream.Dispose() to mark the stream as EOF, or introduce a separate (internal) method for this purpose?

**Comment by hifi** [comment]
@drieseng The `Read()` method is allowed to throw `ObjectDisposedException` by definition so it might be better to set EOF with an internal method.

**Comment by hifi** [comment]
@drieseng Would an empty write do it? A write with zero count would queue normally but when the reader hits that empty buffer it would also return and not dequeue that from the linked list anymore?

**Comment by hifi** [comment]
Actually I had a discussion about Streams on IRC and my eyes were opened that on C# a Stream instance should only do one direction at a time. A pipe should consist of two streams where closing the writer stream also indicates EOF. We should do that instead.

**Comment by drieseng** [comment]
Yeah, we could go for a PipeInputStream/PipeOutputStream approach.
I don't like the fact that in .NET a stream exposes both read and write members.

**Comment by hifi** [comment]
https://gist.github.com/hifi/3882f07f9064ce87e1a43d27a83888b3

I haven't even run this once and it might be overkill and I might have refactored too much but in this you create a `Pipe` which has two public properties `InputStream` and `OutputStream`. You can then decide who to give which and they should only allow you to do the thing you're supposed to be doing and closing `InputStream` would signal EOF.

Tried to share as much code as possible between the streams.

**Comment by drieseng** [comment]
I think I'd prefer to split **PipeStream** into **PipeInputStream** and **PipeOutputStream**, and maybe have **PipeInputStream** "read" from **PipeOutputStream** (not using the ```Stream.Read(byte[] buffer, int offset, int count)``` method on **PipeOutputStream** of course)

**Comment by drieseng** [comment]
Want me to propose something? I think I'll have time tomorrow or Friday.

**Comment by hifi** [comment]
You got the streams mixed but sure, propose something so we have things to compare. One design should stick out to be the cleanest and easiest to understand.

I thought about having the pipe split into two streams but the "ownership" of the data was something that seemed odd to me if one of them had some sort of authority. That's why I used a nested class to hold the state/data and shared that between the two pipes. They could be split into two different classes instead of having a single class that does one of them depending on how it's initialized.

I'll keep experimenting with the design as well.

**Comment by drieseng** [comment]
I don't think I mixed the streams, but feel free to correct me.
Those are the named using in Java.
An **InputStream** in Java is actually a read-only stream.

I won't have time to work on this today :(

**Comment by hifi** [comment]
So it was me who got them mixed up. I'm looking into this today to get those separated and *properly* named.

**Comment by hifi** [comment]
Ok, I've now pushed a new commit to this pull request which replaces all previous ones and it does as follows:

 - PipeStream has been split into Pipe, PipeInputStream and PipeOutputStream
 - Implementation is based on your linked list, although simplified
     - API mostly mimics BlockingCollection
     - Generic\<T\>
 - Because the linked list buffers all data, original behavior of SshCommand should be in place
 - SshCommand has been changed to use this new implementation
 - Closing the other end of the stream (writer) properly indicates EOF for the reader

There's no stdin support in this commit yet. The performance (CPU time) seems to be quite stable when moving large amounts of data (30% @ 1Gbit/s of one core on i6700K on Mono).

Thoughts about this version?

File: src/Renci.SshNet/Common/PipeInputStream.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        public override bool CanRead
+        {
+            get { return true; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
```
Comment (drieseng) [inline src/Renci.SshNet/Common/PipeInputStream.cs:83]: This needs to return true if not disposed.

File: src/Renci.SshNet/Common/PipeOutputStream.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
```
Comment (drieseng) [inline src/Renci.SshNet/Common/PipeOutputStream.cs:27]: If we want to implement this, then the PipeOutputStream needs to know the current position of the PipeInputStream to reject reducing the length to a value less than the current position.

File: src/Renci.SshNet/Common/PipeOutputStream.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            if (buffer == null)
+                throw new ArgumentNullException("buffer");
+            if (offset + count > buffer.Length)
+                throw new ArgumentException("The sum of offset and count is greater than the buffer length.");
+            if (offset < 0 || count < 0)
+                throw new ArgumentOutOfRangeException("offset", "offset or count is negative.");
+            if (_isDisposed)
+                throw CreateObjectDisposedException();
+            if (_queue.IsAddingCompleted)
```
Comment (drieseng) [inline src/Renci.SshNet/Common/PipeOutputStream.cs]: I prefer not to "fail" silently.

File: src/Renci.SshNet/Common/PipeOutputStream.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        public override bool CanWrite
+        {
+            get { return true; }
+        }
+
+        public override long Length
+        {
+            get
+            {
+                throw new NotSupportedException();
```
Comment (drieseng) [inline src/Renci.SshNet/Common/PipeOutputStream.cs:70]: Should be easy to implement.

File: src/Renci.SshNet/Common/Pipe.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        /// <summary>
+        /// Gets the output stream.
+        /// </summary>
+        /// <value>The output stream.</value>
+        public Stream OutputStream { get; private set; }
+
+        public Pipe()
+        {
+            _queue = new LinkedListQueue<byte[]>();
+            InputStream = new PipeInputStream(_queue);
```
Comment (drieseng) [inline src/Renci.SshNet/Common/Pipe.cs:28]: I think we should at least lazily initialize the inputstream.
In that case, we can leave it up to the consumer to dispose it.
In that case, what should PipeInputStream.Read(...) do when the output stream is disposed?
Should it just return 0 signaling EOS?

File: src/Renci.SshNet/Common/Pipe.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        /// </summary>
+        /// <remarks>Call <see cref="Dispose"/> when you are finished using the <see cref="T:Renci.SshNet.Common.Pipe"/>. The
+        /// <see cref="Dispose"/> method leaves the <see cref="T:Renci.SshNet.Common.Pipe"/> in an unusable state. After
+        /// calling <see cref="Dispose"/>, you must release all references to the
+        /// <see cref="T:Renci.SshNet.Common.Pipe"/> so the garbage collector can reclaim the memory that the
+        /// <see cref="T:Renci.SshNet.Common.Pipe"/> was occupying.</remarks>
+        public void Dispose()
+        {
+            OutputStream.Dispose();
+            InputStream.Dispose();
```
Comment (drieseng) [inline src/Renci.SshNet/Common/Pipe.cs:43]: We may not want to implement IDisposable on Pipe. See above.

File: src/Renci.SshNet/Common/Pipe.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
+﻿namespace Renci.SshNet.Common
+{
+    using System;
+    using System.IO;
+
+    /// <summary>
+    /// A generic pipe to pass through data.
+    /// </summary>
+    internal class Pipe : IDisposable
+    {
```
[the remark this answers is not in the document]
Comment (hifi) [reply src/Renci.SshNet/Common/Pipe.cs:10]: Yeah, I kept it for clarity sake for now. Java's `Pipe` also works somewhat like this.

File: src/Renci.SshNet/Common/LinkedListQueue.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
+﻿namespace Renci.SshNet.Common
+{
+    using System;
+    using System.Threading;
+
+    /// <summary>
+    /// Fast concurrent generic linked list queue.
+    /// </summary>
+    public class LinkedListQueue<T> : IDisposable
```
[the remark this answers is not in the document]
Comment (hifi) [reply src/Renci.SshNet/Common/LinkedListQueue.cs]: True, missed that.

File: src/Renci.SshNet/Common/Pipe.cs (hunk 5 above)
Comment (hifi) [reply src/Renci.SshNet/Common/Pipe.cs:28]: What do you mean inintialize in this case? Disposing has been left to "anyone who cares" and multiple disposes aren't an error in this case.

Currently PipeInputStream.Read() will return 0 if the output stream is disposed as it will also finalize writing. Do you think it should be an error condition that the output stream is disposed before input stream has finished reading?

File: src/Renci.SshNet/Common/PipeInputStream.cs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+            if (_current == null || _currentPosition == _current.Length)
+            {
+                if (_queue.IsCompleted || !_queue.TryTake(out _current))
+                    return 0;
+
+                _currentPosition = 0;
+            }
+
+            var avail = _current.Length - _currentPosition;
```
[the remark this answers is not in the document]
Comment (hifi) [reply src/Renci.SshNet/Common/PipeInputStream.cs]: It's still up to Stream specification as it blocks until *some* bytes are available and it will return 0 on EOS. Can be reverted back to a while() loop if it would be better in your opinion to return the full requested buffer size until EOS.

File: src/Renci.SshNet/Common/PipeInputStream.cs (hunk 1 above)
Comment (hifi) [reply src/Renci.SshNet/Common/PipeInputStream.cs:83]: Correct, I didn't notice it's in the Stream spec to return false if closed/disposed.

File: src/Renci.SshNet/Common/PipeOutputStream.cs (hunk 3 above)
Comment (hifi) [reply src/Renci.SshNet/Common/PipeOutputStream.cs]: Should it throw an IOException instead that the stream is closed?

**Comment by hifi** [comment]
FYI I intentionally left out anything strictly not necessary at this point. Meaning seeking, position etc. that you already implemented in the alternative implementation.

Going to fix the small things now and add a new commit for it. Can be rebased later.

File: src/Renci.SshNet/Common/Pipe.cs (hunk 5 above)
Comment (drieseng) [reply src/Renci.SshNet/Common/Pipe.cs:28]: The questions I had were:

- should we immediately create the PipeInputStream, or postpone this until a client is actually requesting one (eg. postpone creating the instance until a client invokes SshCommand.OutputStream)?
- when we postpone creating the PipeInputStream until a client requests it (client could be SSH.NET itself), then shouldn't that client be responsible for disposing this stream (eg. the stream returned by SshCommand.OutputStream)?
- should we have a way to signal write errors to the reader? eg. do we just let PipeInputStream.Read() return 0 when the SSH session has been closed unexpectedly? one of such errors could also be that the writer was disposed before it signaled EOS.

To me, multiple disposes should never be an error. However In this case, that's not what I was trying to say.
See questions above.

File: src/Renci.SshNet/Common/PipeInputStream.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (drieseng) [reply src/Renci.SshNet/Common/PipeInputStream.cs]: We should at least attempt to read up to the specified number of bytes (*count*).
Right now, you only check the current buffer (*_current*) for available bytes.
When the writer has not completed, we should continue reading as long as we have not reached the requested number of bytes, and more data is available.

[the remark this answers is not in the document]
Comment (hifi) [reply src/Renci.SshNet/Common/PipeInputStream.cs]: It could read as much as there's immediately available which would not cause it to block and immediately return if it would block but some data has been read. It would then allow the reader to process the incoming data as fast as it comes in regardless of the read buffer size.

I'll make that change.

File: src/Renci.SshNet/Common/Pipe.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (drieseng) [reply src/Renci.SshNet/Common/Pipe.cs:10]: In Java, the PipeInputStream takes a PipeOutputStream as argument.
There's no **Pipe** class.
For example:
http://tutorials.jenkov.com/java-io/pipes.html

File: src/Renci.SshNet/Common/Pipe.cs (hunk 5 above)
Comment (hifi) [reply src/Renci.SshNet/Common/Pipe.cs:28]: Creating the pipe makes sense as the `Result` property and `Error` property both will read from them when used. Second If you don't create the streams where the incoming data from the server would be buffered if there's output or errors from the executed ssh command?

I think in a normal POSIX setup the streams will just be cut off and the process error code is used to determine if they were complete or not. Correct me if not.

File: src/Renci.SshNet/Common/Pipe.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (hifi) [reply src/Renci.SshNet/Common/Pipe.cs:10]: https://docs.oracle.com/javase/7/docs/api/java/nio/channels/Pipe.html

File: src/Renci.SshNet/Common/Pipe.cs (hunk 5 above)
Comment (drieseng) [reply src/Renci.SshNet/Common/Pipe.cs:28]: We should create the PipeOutputStream, and of course start writing to it.
I don't see why we should immediately create the PipeInputStream (when we don't know yet if anyone will be reading from it), unless a client requests it (not that this client may be us).

File: src/Renci.SshNet/Common/Pipe.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (drieseng) [reply src/Renci.SshNet/Common/Pipe.cs:10]: As long as we don't leak this publicly, I don't mind. However, I still think the actual belongs in the **PipeOutputStream**, and this **PipeOutputStream** should be the only think that is passed to the **PipeInputStream** ctor.

If - big **IF** - we do not want to dispose the **Pipe** by itself (meaning we may want to be able to dispose the writer without disposing the reader), then **Pipe** becomes an empty box.

File: src/Renci.SshNet/Common/Pipe.cs (hunk 5 above)
Comment (hifi) [reply src/Renci.SshNet/Common/Pipe.cs:28]: If you like the non-NIO Java model better, I can refactor this into that and make input stream creation on-demand.

File: src/Renci.SshNet/Common/Pipe.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (hifi) [reply src/Renci.SshNet/Common/Pipe.cs:10]: With non-NIO Java model **Pipe** will become obsolete so if that's a refactor you want I can very well throw things around for that.

[the remark this answers is not in the document]
Comment (drieseng) [reply src/Renci.SshNet/Common/Pipe.cs:10]: I don't necessary want to think in term of what **I** want.
I only think that *if* we decide that we want to lazily initialize the **PipeInputStream** - and only have it disposed by the consumer - then there may not be much interest in have a **Pipe**.

Problem I see is that ```SshCommand.EndExecute(IAsyncResult asyncResult)``` and ```SshCommand.Execute()``` would instantiate the **PipeInputStream** (as they invoke ```SshCommand.Result``` which in turn would lazily instantiate the **PipeInputStream**), but do not directly expose the created **PipeInputStream** to the consumer.  Meaning, the consumer would not be aware that a **PipeInputStream** is created, and that he would need (well *need* is a big word) to dispose it (through ```SshCommand.OutputStream```).

A solution for this would be to modify both ```SshCommand.Result``` and ```SshCommand.Error``` to dispose the **PipeInputStream** if they are creating it.

```cs
public string Result
{
    get
    {
        if (_result == null)
        {
            _result = new StringBuilder();
        }

        if (_outputReadStream != null)
        {
            if (_outputReadStream.DataAvailable)
            {
                // do not dispose the StreamReader, as it would also dispose the stream
                // and we are not the owner of this stream
                var sr = new StreamReader(_outputReadStream, _encoding);
                _result.Append(sr.ReadToEnd());
            }
        }
        else if (_outputWriteStream != null && _outputWriteStream.DataAvailable)
        {
            using (var outputReadStream = new PipeInputStream(_outputWriteStream))
            using (var sr = new StreamReader(outputReadStream, _encoding))
            {
                _result.Append(sr.ReadToEnd());
            }

        }

        return _result.ToString();
    }
}
```

Remarks/questions here:
- Do we want **Result** and **Error** to block until the command has finished execution?
  Currently, they do not block. What do users expect?
- If we do not want these to block, then we need a **DataAvailable** property that returns a value indicating whether there's data available on the stream to be read. (similar to ```NetworkStream.DataAvailable```)

[the remark this answers is not in the document]
Comment (hifi) [reply src/Renci.SshNet/Common/Pipe.cs:10]: What if a result object is returned at the end of execution instead of the properties? In general it doesn't sound like a good idea to have them exposed until execution has ended with any other mean than through the streams.

On way could be that `Execute` would be a shortcut that reads both pipes to EOS and returns a result object with both where the `BeginExecute` and `EndExecute` way could only expose the streams?

[the remark this answers is not in the document]
Comment (drieseng) [reply src/Renci.SshNet/Common/Pipe.cs:10]: Personally, I prefer to remove both properties. However, I want to limit the number of breaking changes.  We could start by marking them as obsolete in the next release (after 2016.1.0).

For now, I guess both properties must be able to return the incomplete output (up until that moment in time).

This means we cannot use ```StreamReader.ReadToEnd()``` as that would block until the command has finished (the writer has completed).

Guess we'll have to do something like this:

```cs
var buffer = new byte[4096];
while (_outputReadStream.DataAvailable)
{
    var bytesRead = _outputReadStream.Read(buffer, 0, buffer.Length);
    _result.Append(_encoding.GetString(buffer, 0, bytesRead));
}
```

[the remark this answers is not in the document]
Comment (hifi) [reply src/Renci.SshNet/Common/Pipe.cs:10]: Then we should never block reading if there's no data available to fill the whole buffer but return with the amount we do have (except block if none) like I've agreed to refactor the reading to include a while loop again.

Though, what's the verdict on the final design of the stream? One thing I did realize is that if you pass the output stream to input stream constructor, they would need to have some shared private API to pass the data which shouldn't be exposed even internally to any other part of SSH.NET. Current implementation avoids that by having a shared data structure - the queue.

This would be all much easier to discuss in a closed room with a whiteboard and beverages.

File: src/Renci.SshNet/Common/PipeInputStream.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (drieseng) [reply src/Renci.SshNet/Common/PipeInputStream.cs]: I agree.

[the remark this answers is not in the document]
Comment (hifi) [reply src/Renci.SshNet/Common/PipeInputStream.cs]: Pushed this change as a new commit.

File: src/Renci.SshNet/Common/Pipe.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (drieseng) [reply src/Renci.SshNet/Common/Pipe.cs:10]: I indeed agree that we should only block until at least some data is available or the writer has signaled EOS.

Was offline for a few hours to install outdoor speakers. Why do this in the summer if you can do it while it's raining and freezing? :p

The **PipeOutputStream** can share the queue with the **PipeInputStream**.
The details can be discussed as we proceed.
Since we only expose **Stream** publicly, we can easily modify our internal design.

[the remark this answers is not in the document]
Comment (drieseng) [reply src/Renci.SshNet/Common/Pipe.cs:10]: I have plently of other stuff I'd like to discuss with you.
The good news is that we've got loads of great beers in Belgium :p

**Comment by hifi** [comment]
Since SshCommand is essentially remote [System.Diagnostics.Process](https://msdn.microsoft.com/en-us/library/system.diagnostics.process(v=vs.110).aspx), why wouldn't we mimic the API as much as we can as it's something people are already used to and it provides all the required things?

Edit: What I mean is that we could instead deprecate SshCommand itself and create SshProcess that's API compatible with Process. This wouldn't break existing use for the next release but provide something you should transition to.

**Comment by 21region** [comment]
Will it be possible to emulate the following command with this new functionality?
```
cat file.tar.gz | ssh user@server 'tar -C /path/to/dest/dir -xjf -'
```

This capability is very desired in our project. For now we are planning to implement it ourselves, but it will be great to see that in an official build of SSH.NET.

**Comment by hifi** [comment]
That is one possible use case for this PR with stdin support.

**Comment by drieseng** [comment]
@hifi The little time I had this week, I've been working on issue #100. I'm making good progress there, so I should soon be able to pick up work on this issue.

**Comment by drieseng** [comment]
@hifi Can we move back to using your [PipeBench](https://github.com/hifi/PipeBench) in order to arrive at a stable API that we both feel comfortable with?

I could also use your help on the performance fix I'm preparing for issue #100.
Note that I still need to commit or submit a PR for this.

**Comment by hifi** [comment]
Sure, PipeBench needs a little bit of cleaning to make it compatible with the double stream approach which should be the one we aim for.

**Comment by drieseng** [comment]
@hifi Could you have a look at performance changes I committed for SFTP download?  More specifically, please have a look at **SftpFileReader** and tell me what you would change or how you would've implemented it.

There's still quite some stuff left to do in that class. Mostly around exceptional handling, but more importantly around cancellation (e.g. how to cancel a blocking wait for a semaphore without having a **CancelleationToken**?). I've been thinking about having our proper implementation of **CancellationToken** that we can use when we're targeting frameworks that don't provide this, but I'm sure there are simpler solutions ...

**Comment by alanmcgovern** [comment]
Out of curiosity, what needs to be done to get this PR completed? I love the idea of exposing a `RemoteProcess` type to mimic `System.Diagnostics.Process`. That approach would clean up some of my code nicely!

**Comment by drieseng** [comment]
I think we kind of abandoned the *process* idea, and were discussing the design of the PipeStream.
I proposed to first finish it in @hifi's [PipeBench](https://github.com/hifi/PipeBench) and then integrate it into **SSH.NET**. Since then I mostly worked on other issues.

I first want to release **2016.1.0-beta2** (about time, it seems) before I continue working on this issue, but I'd gladly welcome you to join the discussion (and - if possible - help with the implementation).

**Comment by hifi** [comment]
Did we abandon the process idea? I just proposed it a few messages up. It did intrigue me to make it look like a remote `System.Diagnostics.Process` if it makes sense to people.

I haven't been active around this project for a few weeks now but I'll get back to it.

**Comment by drieseng** [comment]
"Abondon" was perhaps a bit too strong worded. I though we discussed it, and decided to at least postpone it (and instead first focus on the **PipeStream** design and implementation). But I can't find any trace of it, so I'm probably mistaken. My bad.

Personally, I'd still prefer to move forward like this:
* Design and implement the **PipeStream** (including unit tests)
* Analyze the impact of replacing the current **PipeStream** implementation.
* If this would cause all too many breaking changes, then we MUST introduce a new *remote process* concept.
* If not, we integrate the new implementation into **SSH.NET** and document any change in behavior.

I'd happily discuss and design the *remote process* concept (even if the new **PipeStream** implementation does not introduce too many breaking change), but I know my time is limited (even though I spend already way too much time on **SSH.NET**).

If you guys have enough time available, then please create a new issue for this., I'll participate in the discussions on a best-effort basis.

**Comment by mloefler** [comment]
All the above seems to be quite demanding (timewise).
So I have some quick suggestions:
- Make OutputStream protected in SshCommand since at the end of the execution it is always empty
- Add an OnData event to SshCommand so I can implement my own streaming, optionally avoiding the filling of the OutpuStream, so the memory consumption is not an issue any more.

(Sorry, new here, but just wasted quite some time figuring out why OutputStream does never return anything :-) )

**Comment by hifi** [comment]
I kind of dropped the ball regarding this PR and went into other things. I'd still like to get it going but need to find the motivation again. Regardless the performance is excellent and could even surpass SFTP reading the performance issues it had.

**Comment by drieseng** [comment]
I'd also like this and other PRs to get going again, but I have too little time and would like more interaction/discussion with other devs (eg. on issue #194).

**Comment by mloefler** [comment]
I would not be able to help on those issue, but I will try to do two things (since I need them anyway):
- Add an OnData event to SshCommand, which triggers from inside Channel_DataReceived. This enables me to implement my own data handling and/ or streaming.
- Add a class to 'stream' lines based on input from OnData.
- Modify getter to OutputStream to throw if stream has already been consumed.

I can make the changes available.
(Background: My ssh command extracts large sets of data that I need to process. Large meaning possibly causing a memory issue if all of it resides in memory).

**Comment by thenadz** [comment]
All, any word on this PR? A working OutputStream for the SshCommand would be really nice. Right now I've hacked together some code wrapping the OutputStream to force necessary blocking for a temporary workaround on my end, but if this PR is too large of a scope to get out in the near future, would you consider putting something out just to address this smaller chunk of needed (and easily addressed) functionality.

**Comment by WalkerNA** [comment]
Any progress on this PR?  Our project has a desire that mimic's 21region's example code:

`cat file.tar.gz | ssh user@server 'tar -C /path/to/dest/dir -xjf -'`

**Comment by hifi** [comment]
I think the easiest way is to create something that has a similar API to `System.Diagnostics.Process` and give it a spin like that instead of trying to fix the old `SshCommand`. The old class could then be slowly deprecated and there wouldn't be need to keep compatibility.

Unit testing against said `SshProcess`(?) class could be done by comparing local execution with `Process`.

I haven't made any progress towards this yet so it's open to be picked up as long as there's a pre-accepted solution so no one would be doing more non-mergeable work than I already did.

**Comment by masaeedu** [comment]
I don't know if this is going to work, but I'm cornered enough that I'm now going to attempt the following:

 - Create 3 temp files on the remote system
 - Use `UploadFile` and `DownloadFile` with SFTP to obtain input/output streams that actually work in a sensible way
 - Use `SSHCommand` to start whatever I actually wanted to run under a shell instance with I/O file descriptors redirected to these files. Totally ignore any streams `SSHCommand` provides

Wish me luck 😄

**Comment by ghjkl799** [comment]
> 
> 
> Any progress on this PR? Our project has a desire that mimic's 21region's example code:
> 
> `cat file.tar.gz | ssh user@server 'tar -C /path/to/dest/dir -xjf -'`

Hi

Looking for the same.
Any update on progress?

**Comment by elgonzo** [comment]
Any progress on this would be most appreciated.

Just recently, i was trying to utilize blocking reads with SSH.NET's output `PipeStream`, but it seems the logic associated with the `PipeStream.BlockLastReadBuffer` property is either unintentional/broken or not really useful. If i want to have a blocking read of `X` bytes, why would the `PipeStream.ReadAvailable(int count)` method insist on blocking until `X+1` bytes have been received?

I am not sure if it would make sense to file an issue reportwith regard to my issue (under the condition that i cannot find an equivalent issue report, of course), considering that the issue would probably disappear (or in the worst case morph into a different issue) if the changes discussed here would be implemented in SSH.NET...

**Comment by Livius90** [comment]
When it will be available? I would like to send disk image datas from C# to a stdin pipe of dd over SSH, like in [this example](https://www.thegeekdiary.com/how-to-clone-linux-disk-partition-over-network-using-dd/).

I would like to do the following same thing which works in Linux.
```console
# dd bs=16M if=/dev/sda | ssh root@serverB "dd bs=16M of=/dev/sda"
```

**Comment by WojciechNagorski** [comment]
There is a fork that has this feature 
https://www.nuget.org/packages/TrapTech.SSH.NET
https://github.com/TrapTech/SSH.NET/
But I wish I could use official package.

@Livius90

**Comment by Rob-Hague** [comment]
`SshCommand.CreateInputStream` was added in version 2024.0.0 in #1293 (https://sshnet.github.io/SSH.NET/examples.html#stream-data-to-a-command) and `PipeStream` was rewritten in #1399. I'll put this one to rest.
